src/miniscript/types/extra_props.rs

```diff
@@ -8,6 +8,90 @@ use MiniscriptKey;
 use Terminal;
 
 pub const MAX_OPS_PER_SCRIPT: usize = 201;
+// https://github.com/bitcoin/bitcoin/blob/9ccaee1d5e2e4b79b0a7c29aadb41b97e4741332/src/script/script.h#L39
+pub const HEIGHT_TIME_THRESHOLD: u32 = 500_000_000;
+
+/// Helper struct Whether any satisfaction of this fragment contains any timelocks
+#[derive(Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Debug, Hash)]
+pub struct TimeLockInfo {
+    // csv with heights
+    pub csv_with_height: bool,
+    /// csv with times
+    pub csv_with_time: bool,
+    /// cltv with heights
+    pub cltv_with_height: bool,
+    /// cltv with times
+    pub cltv_with_time: bool,
+    /// combination of any heightlocks and timelocks
+    pub contains_combination: bool,
+}
+
+impl Default for TimeLockInfo {
+    fn default() -> Self {
+        Self {
+            csv_with_height: false,
+            csv_with_time: false,
+            cltv_with_height: false,
+            cltv_with_time: false,
+            contains_combination: false,
+        }
+    }
+}
+
+impl TimeLockInfo {
+    /// Whether the current contains any possible unspendable
+    /// path
+    pub fn contains_unspendable_path(self) -> bool {
+        self.contains_combination
+    }
+
+    // handy function for combining `and` timelocks
+    // This can be operator overloaded in future
+    pub(crate) fn comb_and_timelocks(a: Self, b: Self) -> Self {
+        Self::combine_thresh_timelocks(2, vec![a, b].into_iter())
```

Comment:
Can you do `once(a).chain(once(b))` here rather than `Vec::into_iter`. (You need to `use std::iter::once`.) Saves the allocation.